oauthenticator/google.py

```diff
@@ -373,13 +376,51 @@ def _fetch_user_groups(self, user_email, user_email_domain, http=None):
             credentials=credentials,
             http=http,
         )
+        return service
+
+    def _fetch_member_groups(
+        self,
+        member_email,
+        user_email_domain,
+        http=None,
+        checked_groups=set(),
+        processed_groups=set(),
+    ):
+        """
+        Return a set with the google groups a given user/group is a member of, including nested groups if allowed.
+        """
+        # FIXME: When this function is used and waiting for web request
+        #        responses, JupyterHub gets blocked from doing other things.
+        #        Ideally the web requests should be made using an async client
+        #        that can be awaited while JupyterHub handles other things.
+        #
+        if not hasattr(self, 'service'):
+            self.service = self._setup_service(user_email_domain, http)
 
-        resp = service.groups().list(userKey=user_email).execute()
-        user_groups = {
-            g['email'].split('@')[0] for g in resp.get('groups', [{'email': None}])
+        resp = self.service.groups().list(userKey=member_email).execute()
+        member_groups = {
+            g['email'].split('@')[0] for g in resp.get('groups', []) if g.get('email')
         }
-        self.log.debug(f"user_email {user_email} is a member of {user_groups}")
-        return user_groups
+        self.log.debug(f"Fetched groups for {member_email}: {member_groups}")
+
+        checked_groups.update(member_groups)
+        self.log.debug(f"Checked groups after update: {checked_groups}")
+
+        if self.allow_nested_groups:
+            for group in member_groups:
+                if group not in processed_groups:
+                    processed_groups.add(group)
+                    nested_groups = self._fetch_member_groups(
+                        f"{group}@{user_email_domain}",
+                        user_email_domain,
+                        http,
+                        checked_groups,
+                        processed_groups,
+                    )
+                    checked_groups.update(nested_groups)
```

Comment:
I've come to appreciate de-nesting code like this, but its just an opinionated refactor we don't need to do.

```suggestion
                if group in processed_groups:
                    continue
                processed_groups.add(group)
                nested_groups = self._fetch_member_groups(
                    f"{group}@{user_email_domain}",
                    user_email_domain,
                    http,
                    checked_groups,
                    processed_groups,
                )
                checked_groups.update(nested_groups)
```